packages/medusa/src/services/product.ts

```diff
@@ -400,112 +420,141 @@ class ProductService extends TransactionBaseService<ProductService> {
-        this.productVariantRepository_
-      )
-      const productTagRepo = manager.getCustomRepository(
-        this.productTagRepository_
-      )
-      const productTypeRepo = manager.getCustomRepository(
-        this.productTypeRepository_
-      )
-      const imageRepo = manager.getCustomRepository(this.imageRepository_)
-
-      const product = await this.retrieve(productId, {
-        relations: ["variants", "tags", "images"],
-      })
+    return await this.atomicPhase_(
+      async (manager) => {
+        const productRepo = manager.getCustomRepository(this.productRepository_)
+        const productVariantRepo = manager.getCustomRepository(
+          this.productVariantRepository_
+        )
+        const productTagRepo = manager.getCustomRepository(
+          this.productTagRepository_
+        )
+        const productTypeRepo = manager.getCustomRepository(
+          this.productTypeRepository_
+        )
+        const imageRepo = manager.getCustomRepository(this.imageRepository_)
 
-      const { variants, metadata, images, tags, type, ...rest } = update
+        const product = await this.retrieve(productId, {
+          relations: ["variants", "tags", "images", "sales_channels"],
+        })
 
-      if (!product.thumbnail && !update.thumbnail && images?.length) {
-        product.thumbnail = images[0]
-      }
+        const {
+          variants,
+          metadata,
+          images,
+          tags,
+          type,
+          sales_channels: salesChannels,
+          ...rest
+        } = update
+
+        if (!product.thumbnail && !update.thumbnail && images?.length) {
+          product.thumbnail = images[0]
+        }
 
-      if (images) {
-        product.images = await imageRepo.upsertImages(images)
-      }
+        if (images) {
+          product.images = await imageRepo.upsertImages(images)
+        }
 
-      if (metadata) {
-        product.metadata = setMetadata(product, metadata)
-      }
+        if (metadata) {
+          product.metadata = setMetadata(product, metadata)
+        }
 
-      if (typeof type !== `undefined`) {
-        product.type_id = (await productTypeRepo.upsertType(type))?.id || null
-      }
+        if (typeof type !== `undefined`) {
+          product.type_id = (await productTypeRepo.upsertType(type))?.id || null
+        }
 
-      if (tags) {
-        product.tags = await productTagRepo.upsertTags(tags)
-      }
+        if (tags) {
+          product.tags = await productTagRepo.upsertTags(tags)
+        }
 
-      if (variants) {
-        // Iterate product variants and update their properties accordingly
-        for (const variant of product.variants) {
-          const exists = variants.find((v) => v.id && variant.id === v.id)
-          if (!exists) {
-            await productVariantRepo.remove(variant)
-          }
+        if (typeof salesChannels !== "undefined") {
+          product.sales_channels =
+            salesChannels === null || salesChannels?.length === 0
```

Comment:
nit(non-blocking): you can just check `salesChannels?.length` and invert the ternary since `null?.length` is `undefined`